File: codes/matlab/ch05/fd_matrix_bandwidth.m
%% fd_matrix_bandwidth.m
%
% Visualizes the sparsity patterns of finite difference and spectral
% differentiation matrices, showing how the bandwidth expands as the
% order of accuracy increases.
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
% Part of "Computational Études: A Spectral Approach"
% https://github.com/dutykh/computational-etudes

clear; close all; clc;

%% Configuration
N = 20;  % Number of grid points

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch05', 'matlab');
output_file = fullfile(output_dir, 'fd_matrix_bandwidth.pdf');

% Create output directory if needed
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

%% Construct matrices
[D_fd2, ~] = fd_matrix_periodic(N, 2);   % 2nd order: tridiagonal
[D_fd4, ~] = fd_matrix_periodic(N, 4);   % 4th order: pentadiagonal
[D_spec, ~] = spectral_matrix_periodic(N);  % Spectral: dense

%% Create figure
fig = figure('Units', 'inches', 'Position', [1, 1, 11, 3.5]);

matrices = {D_fd2, D_fd4, D_spec};
titles = {'2nd Order FD', '4th Order FD', 'Spectral'};
bandwidths = {'Bandwidth = 3', 'Bandwidth = 5', sprintf('Dense (N \\times N)')};

for k = 1:3
    subplot(1, 3, k);

    % Create binary sparsity pattern
    threshold = 1e-14;
    sparsity = abs(matrices{k}) > threshold;

    % Plot using imagesc
    imagesc(sparsity);
    colormap([1 1 1; 0.078 0.176 0.431]);  % White and Navy

    title(titles{k}, 'FontSize', 12, 'FontWeight', 'bold');
    % Combine column label and bandwidth info
    xlabel({['Column index ' char(106)], bandwidths{k}}, 'FontSize', 10);
    if k == 1
        ylabel('Row index i', 'FontSize', 10);
    end

    axis square;
    set(gca, 'XTick', [1, N/2, N], 'XTickLabel', {'0', num2str(N/2-1), num2str(N-1)});
    set(gca, 'YTick', [1, N/2, N], 'YTickLabel', {'0', num2str(N/2-1), num2str(N-1)});
end

sgtitle('Finite Difference Matrix Sparsity: From Local to Global', 'FontSize', 13);

%% Save figure
exportgraphics(fig, output_file, 'ContentType', 'vector');
exportgraphics(fig, strrep(output_file, '.pdf', '.png'), 'Resolution', 300);

fprintf('Figure saved to: %s\n', output_file);

%% Print statistics
fprintf('\nMatrix Statistics (N = %d):\n', N);
fprintf('--------------------------------------------------\n');
for k = 1:3
    nnz_count = sum(abs(matrices{k}(:)) > 1e-14);
    density = nnz_count / numel(matrices{k}) * 100;
    fprintf('%15s: %4d nonzeros (%5.1f%% dense)\n', titles{k}, nnz_count, density);
end
fprintf('--------------------------------------------------\n');

close(fig);

File: codes/julia/ch05/fd_matrix_bandwidth.jl
# fd_matrix_bandwidth.jl
#
# Visualizes the sparsity patterns of finite difference and spectral
# differentiation matrices, showing how the bandwidth expands as the
# order of accuracy increases.
#
# Key Message:
#     - 2nd order FD: Tridiagonal (bandwidth 3)
#     - 4th order FD: Pentadiagonal (bandwidth 5)
#     - Spectral: Dense (bandwidth N)
#
#     This illustrates that spectral methods are the limit case of finite
#     differences as the stencil width extends to the full domain.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

# Include shared utilities
include(joinpath(@__DIR__, "fdweights.jl"))

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const NAVY  = colorant"#142D6E"
const SKY   = colorant"#7896D2"
const CORAL = colorant"#E74C3C"
const TEAL  = colorant"#16A085"

# Output path
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch05", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Periodic spectral differentiation matrix
# -----------------------------------------------------------------------------
"""
    spectral_diff_periodic(N)

Construct the periodic spectral differentiation matrix for N equispaced
points on [0, 2pi).

Returns `(D, x)` where D is the N x N differentiation matrix and x are
the grid points.
"""
function spectral_diff_periodic(N)
    N < 2 && error("Need at least 2 points")

    h = 2pi / N
    x = h .* (0:N-1)

    D = zeros(N, N)
    for i in 1:N
        for j in 1:N
            if i != j
                diff = i - j
                D[i, j] = 0.5 * ((-1)^diff) / tan(diff * pi / N)
            end
        end
    end

    return D, x
end

# -----------------------------------------------------------------------------
# Finite difference matrix construction
# -----------------------------------------------------------------------------
"""
    fd_diff_periodic(N, order)

Construct a periodic finite difference differentiation matrix.
"""
function fd_diff_periodic(N, order)
    h = 2pi / N

    # Determine stencil size
    stencil_half = order ÷ 2

    # Create stencil nodes centered at 0
    stencil_nodes = h .* (-stencil_half:stencil_half)

    # Compute FD weights for first derivative at center
    weights = fdweights(0.0, collect(stencil_nodes), 1)

    # Build circulant matrix
    D = zeros(N, N)
    for i in 1:N
        for (k, w) in enumerate(weights)
            j = mod(i - 1 + k - 1 - stencil_half, N) + 1  # Periodic wrap-around (1-indexed)
            D[i, j] += w
        end
    end

    return D
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    N = 20  # Number of grid points

    # Create the three matrices
    D_fd2 = fd_diff_periodic(N, 2)   # 2nd order: tridiagonal
    D_fd4 = fd_diff_periodic(N, 4)   # 4th order: pentadiagonal
    D_spec, _ = spectral_diff_periodic(N)  # Spectral: dense

    # Create figure with three panels
    fig = Figure(size = (880, 300))

    matrices = [D_fd2, D_fd4, D_spec]
    titles = ["2nd Order FD", "4th Order FD", "Spectral"]
    bandwidths = ["Bandwidth = 3", "Bandwidth = 5", "Dense (N x N)"]

    for (idx, (D, title, bw)) in enumerate(zip(matrices, titles, bandwidths))
        ax = Axis(fig[1, idx],
                  title = title,
                  titlesize = 12,
                  xlabel = "Column index j\n$bw",
                  xlabelsize = 10,
                  ylabel = idx == 1 ? "Row index i" : "",
                  ylabelsize = 10,
                  yreversed = true,
                  aspect = DataAspect())

        # Create binary sparsity pattern
        threshold = 1e-14
        sparsity = Float64.(abs.(D) .> threshold)

        # Plot using heatmap
        heatmap!(ax, 0:N-1, 0:N-1, sparsity',
                 colormap = [:white, NAVY],
                 colorrange = (0, 1))

        # Reduce tick density
        ax.xticks = [0, N ÷ 2, N - 1]
        ax.yticks = [0, N ÷ 2, N - 1]
    end

    # Add overall title
    Label(fig[0, :], "Finite Difference Matrix Sparsity: From Local to Global",
          fontsize = 13, font = :bold)

    # Save figure
    save(joinpath(OUTPUT_DIR, "fd_matrix_bandwidth.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "fd_matrix_bandwidth.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "fd_matrix_bandwidth.pdf"))")

    # Print matrix statistics
    println("\nMatrix Statistics (N = $N):")
    println("-" ^ 50)
    for (D, title) in zip(matrices, titles)
        nnz = count(x -> abs(x) > 1e-14, D)
        density = nnz / length(D) * 100
        println("$(rpad(title, 15)): $(lpad(string(nnz), 4)) nonzeros ($(lpad(@sprintf("%.1f", density), 5))% dense)")
    end
    println("-" ^ 50)
end

main()
